torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -201,6 +201,18 @@ void ProcessGroupNCCL::WorkNCCL::synchronize() {
       auto currentTimepoint = std::chrono::steady_clock::now();
       if (std::chrono::duration_cast<std::chrono::milliseconds>(
               currentTimepoint - workStartTime_) > opTimeout_) {
+        // When operation times out due to some errors that are not
+        // detected by nccl communicators, ncclCommWatchdog can not check this
+        // time out error and thus can not abort ncclComms accordingly.
+        // So explicitly abort ncclComms here before throwing this timed out
+        // exception to users, after this, ncclCommWatchdog can detect nccl
+        // communicators are aborted and clean up devNCCLCommMap_ accordingly.
+        // if throwing timed out excepiton without aborting nccl communicators
+        // here, it was observed that CUDA GPU will have 100% utilization and
+        // can not run new events successfully.
```

Comment:
Did we verify that the watchdog detects an issue with the communicator and successfully removes the communicator? I think we should add a test in `ProcessGroupNCCLErrorsTest` to ensure that the communicators are cleared from the cache.

Reply:
yes, local prints showed watch dog received error and clean it up accordingly. OK, will think about how to verify communicators are cleared

Reply:
added a test for timedout error in ProcessGroupNCCLErrorsTest, it validated that the communicators are cleared from the cache as well